### DISCUSSION TITLE
Constrained ANN search with filtering

### DISCUSSION THREAD

**Discussion by josephsachdeva** [lmcinnes/pynndescent#206] [Ideas]
Is there a way that we can do an ANN search using filters? For example: I want to fetch documents only with language 'english'. This can be added as a tag while building index.

OR

What HNSW provides is, we can mark 'ids' as mark_deleted so the ANN would not search those documents. It will be more useful for pre-filtering the documents.

**Comment by lmcinnes** [comment]
The newer versions of the ``update`` method provide some ability to delete or modify nodes, as well as add them. I suspect however, that you really want constrained search rather than dealing with deletions. This is certainly something that is on my list. In practice what I had in mind is a "masked query" where you provide a query and a mask vector that specifies (via a binary indicator) which items to include in results (or not). This gets tricky in some extreme cases if you want the search to be efficient, and depending on how the filtered items are distributed among the data. It is a feature I would like to hyave, but I don't have any timelines for when it might be available.
